File: backend/chessapp/domain/pieces/piece_factory.py
from ...domain.pieces.bishop import Bishop
from ...domain.pieces.king import King
from ...domain.pieces.knight import Knight
from ...domain.pieces.pawn import Pawn
from ...domain.pieces.piece_type import PieceType
from ...domain.pieces.queen import Queen
from ...domain.pieces.rook import Rook
from ...domain.value_objects.side import Side


class PieceFactory(object):

    @classmethod
    def create(cls, side: Side, type_: PieceType):
        match type_:
            case PieceType.Pawn:
                return Pawn(side)
            case PieceType.King:
                return King(side)
            case PieceType.Rook:
                return Rook(side)
            case PieceType.Queen:
                return Queen(side)
            case PieceType.Bishop:
                return Bishop(side)
            case PieceType.Knight:
                return Knight(side)
        return None

File: frontend/src/app/pages/game/play/play/models/pieces/piece_factory.ts
import { PieceDto } from "src/app/services/models/chess-game-dto";
import { Side } from "../side";
import { Bishop } from "./bishop";
import { King } from "./king";
import { Knight } from "./knight";
import { Pawn } from "./pawn";
import { Queen } from "./queen";
import { Rook } from "./rook";

export class PieceFactory {
    
    static getPiece(piece: PieceDto) {
        switch (piece.abbreviation.toLowerCase()) {
            case 'wp':
                return new Pawn(Side.white);
            case 'bp':
                return new Pawn(Side.black);
            case 'wr':
                return new Rook(Side.white);
            case 'br':
                return new Rook(Side.black);
            case 'wb':
                return new Bishop(Side.white);
            case 'bb':
                return new Bishop(Side.black);
            case 'wn':
                return new Knight(Side.white);
            case 'bn':
                return new Knight(Side.black);
            case 'wq':
                return new Queen(Side.white);
            case 'bq':
                return new Queen(Side.black);
            case 'wk':
                return new King(Side.white);
            case 'bk':
                return new King(Side.black);
            default:
                throw Error('Un-recognized piece type');
        }
    }
}